client/api_client.go

```diff
@@ -0,0 +1,522 @@
+type Option struct {
+	// GrpcDialOption is the list of gRPC dial options used to configure the connection to the blockchain node.
+	GrpcDialOption grpc.DialOption
+	// account used to set the default account of client
+	DefaultAccount *types.Account
+	// Secure is a flag that specifies whether the client should use HTTPS or not.
+	Secure bool
+	// Transport is the HTTP transport used to send requests to the storage provider endpoint.
+	Transport http.RoundTripper
+	// Host is the target sp server hostname
+	Host string
+}
+
+// New - instantiate greenfield chain with chain info, account info and options.
+// The grpcAddress indicate the grpc address of greenfield chain.
+// The account indicate the account used for signing transactions.
+// The rpcAddress indicate the rpc address of the tendermint client.
+func New(chainID string, endpoint string, option Option) (Client, error) {
+	if endpoint == "" || chainID == "" {
+		return nil, errors.New("fail to get grpcAddress and chainID to construct client")
+	}
+	cc, err := sdkclient.NewGreenfieldClient(endpoint, chainID)
+	if err != nil {
+		return nil, err
+	}
+	if option.DefaultAccount != nil {
+		cc.SetKeyManager(option.DefaultAccount.GetKeyManager())
+	}
+
+	c := client{
+		chainClient:    cc,
+		httpClient:     &http.Client{Transport: option.Transport},
+		userAgent:      types.UserAgent,
+		defaultAccount: option.DefaultAccount, // it allows to be nil
+		secure:         option.Secure,
+		host:           option.Host,
+	}
+
+	// fetch sp endpoints info from chain
+	spInfo, err := c.getSPUrlInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	c.spEndpoints = spInfo
+	return &c, nil
+}
+
+// getSPUrlByBucket route url of the sp from bucket name
+func (c *client) getSPUrlByBucket(bucketName string) (*url.URL, error) {
+	ctx := context.Background()
+	bucketInfo, err := c.HeadBucket(ctx, bucketName)
+	if err != nil {
+		return nil, err
+	}
+
+	primarySP := bucketInfo.GetPrimarySpAddress()
+	if _, ok := c.spEndpoints[primarySP]; ok {
+		return c.spEndpoints[primarySP], nil
+	}
+	// query sp info from chain
+	newSpInfo, err := c.getSPUrlInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := newSpInfo[primarySP]; ok {
+		c.spEndpoints = newSpInfo
+		return newSpInfo[primarySP], nil
+	} else {
+		return nil, errors.New("fail to locate endpoint from bucket")
+	}
+}
+
+// getSPUrlByAddr route url of the sp from sp address
+func (c *client) getSPUrlByAddr(address string) (*url.URL, error) {
+	if _, ok := c.spEndpoints[address]; ok {
+		return c.spEndpoints[address], nil
+	}
+	// query sp info from chain
+	newSpInfo, err := c.getSPUrlInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := newSpInfo[address]; ok {
+		c.spEndpoints = newSpInfo
+		return newSpInfo[address], nil
+	} else {
+		return nil, errors.New("fail to locate endpoint from address")
+	}
+}
+
+// getInServiceSP return an SP endpoint which is in service
+func (c *client) getInServiceSP() (*url.URL, error) {
```

Comment:
To get in service SP list? When to use this function?

Reply:
It return the first SP of the SP list which are service.  ListBuckets use it for now.  ListBuckets  have to access meta service of SP,  but it has no method to get the SP address,  so need get a  in service SP for it .   
It can access all sp and get the right result .